### PR TITLE
[FEATURE] Withdraw rewards functionality for agent and manager contracts

### PR DESCRIPTION
**Pull request by deveusss**
See more in #287

### CODE REVIEW

File: contracts/croncat-agents/src/contract.rs [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
 use crate::distributor::*;
 use crate::error::ContractError;
 use crate::external::*;
 use crate::msg::*;
 use crate::state::*;
 #[cfg(not(feature = "library"))]
 use cosmwasm_std::entry_point;
+use cosmwasm_std::CosmosMsg;
+use cosmwasm_std::Empty;
+use cosmwasm_std::QuerierWrapper;
```
Comment (TrevorJTClarke) [inline contracts/croncat-agents/src/contract.rs:10]: @deveusss these deps should all be in the below list

File: contracts/croncat-agents/src/contract.rs [hunk 2]
```diff
@@ -90,21 +92,24 @@ pub fn execute(
...
         ExecuteMsg::UnregisterAgent { from_behind } => {
             unregister_agent(deps, &info.sender, from_behind)
         }
+        ExecuteMsg::UpdateAgent { payable_account_id } => {
+            update_agent(deps, info, env, payable_account_id)
+        }
         ExecuteMsg::CheckInAgent {} => accept_nomination_agent(deps, info, env),
         ExecuteMsg::OnTaskCreated(msg) => on_task_created(env, deps, info, msg),
         ExecuteMsg::UpdateConfig { config } => execute_update_config(deps, info, config),
+        ExecuteMsg::WithdrawRewards {} => execute_withdraw_rewards(deps, info),
```
Comment (TrevorJTClarke) [inline contracts/croncat-agents/src/contract.rs:104]: @deveusss this implementation needs to live within the manager contract with the other balances functionality

File: contracts/croncat-agents/src/contract.rs [hunk 3]
```diff
@@ -607,6 +617,55 @@ fn on_task_created(
             AGENT_NOMINATION_BEGIN_TIME.save(deps.storage, &Some(env.block.time))?;
         }
     }
-    let response = Response::new().add_attribute("method", "on_task_created");
+    let response = Response::new().add_attribute("action", "on_task_created");
+    Ok(response)
+}
+
+pub fn execute_withdraw_rewards(
```
Comment (TrevorJTClarke) [inline contracts/croncat-agents/src/contract.rs:624]: move to manager

File: contracts/croncat-agents/src/contract.rs [hunk 4]
```diff
@@ -607,6 +617,55 @@ fn on_task_created(
...
+
+    let response = Response::new()
+        .add_attribute("action", "withdraw_rewards")
+        .add_attribute("agent_id", agent_id)
+        .add_attribute("balance", balance)
+        .add_messages(msgs);
     Ok(response)
 }
+
+pub(crate) fn withdraw_rewards(
```
Comment (TrevorJTClarke) [inline contracts/croncat-agents/src/contract.rs:641]: move to manager

File: contracts/croncat-agents/src/external.rs [hunk 5]
```diff
@@ -7,78 +8,123 @@ use croncat_sdk_tasks::types::SlotTasksTotalResponse;
...
+            addr,
+            &ManagerQueryMsg::AgentRewards {
+                agent_id: agent_id.to_owned(),
+            },
+        )?;
 
-    Ok((response.block_tasks, response.cron_tasks))
+        Ok(response)
+    }
+    pub fn create_withdraw_rewards_submsg(
```
Comment (TrevorJTClarke) [inline contracts/croncat-agents/src/external.rs:107]: not needed, when moved to manager

File: contracts/croncat-manager/src/contract.rs [hunk 6]
```diff
@@ -351,3 +360,63 @@ pub fn reply(_deps: DepsMut, _env: Env, _msg: Reply) -> Result<Response, Contrac
...
+            balance: agent.balance,
+            payable_account_id: agent.payable_account_id.to_string(),
+        };
+
+        if callback.balance.is_zero() {
+            return Err(ContractError::NoWithdrawRewardsAvailable {});
+        }
+    }
+
+    let rewards = TREASURY_BALANCE.load(deps.storage)?;
```
Comment (TrevorJTClarke) [inline contracts/croncat-manager/src/contract.rs:396]: @deveusss this is actually wrong.
The manager has 2 buckets of balances (which is why this logic needs to move) where we have 1 for agents, 1 for treasury (which only pays the DAO). This logic actually crosses the boundary of 2 diff reward pools

**Review by TrevorJTClarke — CHANGES_REQUESTED** [review]
@deveusss see comments below - this needs logic moved.
Manager is the only contract for balance handling, agent cannot have any logic handling balances.

File: contracts/croncat-manager/src/contract.rs (hunk 6 above)
Comment (TrevorJTClarke) [reply contracts/croncat-manager/src/contract.rs:396]: See here: https://github.com/CronCats/cw-croncat/blob/beta/split-0.0.0-withdraw-rewards/contracts/croncat-manager/src/state.rs#L13

File: contracts/croncat-agents/src/contract.rs (hunk 2 above)
Comment (deveusss) [reply contracts/croncat-agents/src/contract.rs]: √

File: contracts/croncat-agents/src/contract.rs (hunk 3 above)
Comment (deveusss) [reply contracts/croncat-agents/src/contract.rs]: √

File: contracts/croncat-agents/src/external.rs (hunk 5 above)
Comment (deveusss) [reply contracts/croncat-agents/src/external.rs:107]: this one need for agent unregister

File: contracts/croncat-manager/src/contract.rs [hunk 7]
```diff
@@ -370,53 +370,50 @@ fn execute_withdraw_agent_rewards(
...
-            balance: agent.balance,
-            payable_account_id: agent.payable_account_id.to_string(),
-        };
-
-        if callback.balance.is_zero() {
-            return Err(ContractError::NoWithdrawRewardsAvailable {});
-        }
+        payable_account_id = agent.payable_account_id;
     }
+    let agent_rewards = AGENT_REWARDS
```
Comment (TrevorJTClarke) [inline contracts/croncat-manager/src/contract.rs:389]: awesome, thank you!

**Review by TrevorJTClarke — APPROVED** [review]

**Pull request merged**